whois-api/src/test/java/net/ripe/db/whois/api/rest/WhoisRestServiceDoSTestIntegration.java

```diff
@@ -0,0 +1,289 @@
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+
+@Tag("IntegrationTest")
+public class WhoisRestServiceDoSTestIntegration extends AbstractIntegrationTest {
+
+    @Autowired
+    private WhoisObjectMapper whoisObjectMapper;
+
+    private static final RpslObject OWNER_MNT = RpslObject.parse("" +
+            "mntner:      OWNER-MNT\n" +
+            "descr:       Owner Maintainer\n" +
+            "admin-c:     TP1-TEST\n" +
+            "upd-to:      noreply@ripe.net\n" +
+            "auth:        MD5-PW $1$d9fKeTr2$Si7YudNf4rUGmR71n/cqk/ #test\n" +
+            "auth:        SSO person@net.net\n" +
+            "mnt-by:      OWNER-MNT\n" +
+            "source:      TEST");
+
+    @BeforeAll
+    public static void beforeClass() {
+        System.setProperty("dos.filter.enabled", "true");
+    }
+
+    @BeforeEach
+    public void setup() {
+        databaseHelper.addObject("person: Test Person\nnic-hdl: TP1-TEST");
+        databaseHelper.addObject(OWNER_MNT);
+        testDateTimeProvider.setTime(LocalDateTime.parse("2001-02-04T17:00:00"));
+    }
+
+    // LookUps
+    @Test
+    public void multiple_lookUps_per_second_then_429_too_many_requests() throws InterruptedException {
+        final Response response = RestTest.target(getPort(), "whois/test/mntner/OWNER-MNT?clientIp=10.20.30.40")
+                .request()
+                .get();
+        assertThat(HttpStatus.OK_200, is(response.getStatus()));
+
+        final Invocation.Builder lookupRequest = RestTest.target(getPort(), "whois/test/mntner/OWNER-MNT?clientIp=10.20.30.40").request();
+        final List<Integer> statuses = Lists.newArrayList();
+        for (int i = 0; i < 50; i++) {
+            statuses.add(lookupRequest.get().getStatus());
+        }
+
+        assertThat(HttpStatus.OK_200, is(statuses.get(48)));
+        assertThat(HttpStatus.TOO_MANY_REQUESTS_429, is(statuses.getLast()));
+
+        TimeUnit.SECONDS.sleep(1); // Free the IP after one second
+
+        //After a second, the user can perform more requests
+        assertThat(HttpStatus.OK_200, is(lookupRequest.get().getStatus()));
+    }
+
+    @Test
+    public void multiple_async_lookUps_per_second_then_429_too_many_requests() throws InterruptedException {
+        final Response response = RestTest.target(getPort(), "whois/test/mntner/OWNER-MNT?clientIp=10.20.30.40")
+                .request()
+                .get();
+        assertThat(HttpStatus.OK_200, is(response.getStatus()));
+
+        final AsyncInvoker lookupRequest = RestTest.target(getPort(), "whois/test/mntner/OWNER-MNT?clientIp=10.20.30.40").request().async();
+
+        // Simulate a DoS attack by sending many POST requests in a short time asynchronously
+        final List<CompletableFuture<Response>> futures = IntStream.range(0, 50)
```

Comment:
Can you simplify using a `parallel()` stream instead ?

Reply:
Updates using streams and parallel()